### PR TITLE
[Core] Display allowed values in error message when enum validation fails

### PR DESCRIPTION
**Pull request by xfz11**
**Description**

When the enum validation fails, error message would show the allowed values to improve user experience
Before:
image
After:
image

The case in which the candiate list is shown along with the best match candidate:
image

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

File: src/azure-cli-core/azure/cli/core/parser.py [hunk 1]
```diff
@@ -497,8 +497,8 @@ def _check_value(self, action, value):  # pylint: disable=too-many-statements, t
             else:
                 # `command_source` indicates command values have been parsed, value is an argument
                 parameter = action.option_strings[0] if action.option_strings else action.dest
-                error_msg = "{prog}: '{value}' is not a valid value for '{param}'.".format(
-                    prog=self.prog, value=value, param=parameter)
+                error_msg = "{prog}: '{value}' is not a valid value for '{param}'. Allowed values: {choice}.".format(
+                    prog=self.prog, value=value, param=parameter, choice=', '.join(sorted([str(x) for x in action.choices])))
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/parser.py:501]: Do we need `sorted`? Shall we respect the order in swagger definition?

File: src/azure-cli-core/azure/cli/core/parser.py [hunk 2]
```diff
@@ -497,8 +497,8 @@ def _check_value(self, action, value):  # pylint: disable=too-many-statements, t
             else:
                 # `command_source` indicates command values have been parsed, value is an argument
                 parameter = action.option_strings[0] if action.option_strings else action.dest
-                error_msg = "{prog}: '{value}' is not a valid value for '{param}'.".format(
-                    prog=self.prog, value=value, param=parameter)
+                error_msg = "{prog}: '{value}' is not a valid value for '{param}'. Allowed values: {choice}.".format(
```
Comment (qwordy) [inline src/azure-cli-core/azure/cli/core/parser.py:500]:
```suggestion
                error_msg = "{prog}: '{value}' is not a valid value for '{param}'. Allowed values: {choices}.".format(
```

**Comment by houk-ms** [comment]
@jiasli for awareness. I remember we discussed this before, any plans for this feature in Knack?

**Comment by yonzhan** [comment]
@jiasli for awareness

File: src/azure-cli-core/azure/cli/core/parser.py (hunk 1 above)
Comment (xfz11) [reply src/azure-cli-core/azure/cli/core/parser.py:501]: The order in error message is better to be consistent with the order in help info

**Comment by houk-ms** [comment]
@xfz11 Can you help paste some creenshots in PR description about what the chanegs will look like before and after?

The cases we need to consider include
1. When the canidate list is very long
2. When the candiate list is shown along with the best match candidate

File: src/azure-cli-core/azure/cli/core/parser.py (hunk 1 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/parser.py:501]: I personally prefer to keep the order because some choices may be more frequently used.

**Comment by jiasli** [comment]
The knack implementation is https://github.com/microsoft/knack/pull/244 and have been released in [v0.8.1](https://github.com/microsoft/knack/releases/tag/v0.8.1). I have no strong opinion in CLI's behavior and will defer that to @houk-ms.

**Comment by houk-ms** [comment]
@Juliehzl Do you have concerns about it if we decide to display all the possbile candidates?

**Review by houk-ms — APPROVED** [review]

**Pull request merged**